src/bootstrap/bootstrap.py

```diff
@@ -0,0 +1,326 @@
+
+    def cargo_stamp(self):
+        return os.path.join(self.bin_root(), '.cargo-stamp')
+
+    def rustc_out_of_date(self):
+        if not os.path.exists(self.rustc_stamp()):
+            return True
+        with open(self.rustc_stamp(), 'r') as f:
+            return self.snap_rustc_date() != f.read()
+
+    def cargo_out_of_date(self):
+        if not os.path.exists(self.cargo_stamp()):
+            return True
+        with open(self.cargo_stamp(), 'r') as f:
+            return self.snap_cargo_date() != f.read()
+
+    def bin_root(self):
+        return os.path.join(self.build_dir, self.build, "stage0")
+
+    def get_toml(self, key):
+        for line in iter(self.config_toml.splitlines()):
+            if line.startswith(key + ' ='):
+                return self.get_string(line)
+        return None
+
+    def get_mk(self, key):
+        for line in iter(self.config_mk.splitlines()):
+            if line.startswith(key):
+                return line[line.find(':=') + 2:].strip()
+        return None
+
+    def cargo(self):
+        config = self.get_toml('cargo')
+        if config:
+            return config
+        return os.path.join(self.bin_root(), "bin/cargo" + self.exe_suffix())
+
+    def rustc(self):
+        config = self.get_toml('rustc')
+        if config:
+            return config
+        config = self.get_mk('CFG_LOCAL_RUST')
+        if config:
+            return config + '/bin/rustc' + self.exe_suffix()
+        return os.path.join(self.bin_root(), "bin/rustc" + self.exe_suffix())
+
+    def get_string(self, line):
+        start = line.find('"')
+        end = start + 1 + line[start+1:].find('"')
+        return line[start+1:end]
+
+    def exe_suffix(self):
+        if sys.platform == 'win32':
+            return '.exe'
+        else:
+            return ''
+
+    def parse_nightly_dates(self):
+        nightlies = os.path.join(self.rust_root, "src/nightlies.txt")
+        rustc, cargo = open(nightlies, 'r').read().split("\n")[:2]
+        assert(rustc.startswith("rustc: "))
+        assert(cargo.startswith("cargo: "))
+        self._rustc_date = rustc[len("rustc: "):]
+        self._cargo_date = cargo[len("cargo: "):]
+
+    def build_bootstrap(self):
+        env = os.environ.copy()
+        env["CARGO_TARGET_DIR"] = os.path.join(self.build_dir, "bootstrap")
+        env["RUSTC"] = self.rustc()
+        env["LD_LIBRARY_PATH"] = os.path.join(self.bin_root(), "lib")
+        env["DYLD_LIBRARY_PATH"] = os.path.join(self.bin_root(), "lib")
+        env["PATH"] = os.path.join(self.bin_root(), "bin") + \
+                      os.pathsep + env["PATH"]
+        self.run([self.cargo(), "build", "--manifest-path",
+                  os.path.join(self.rust_root, "src/bootstrap/Cargo.toml")],
+                 env)
+
+    def run(self, args, env):
+        proc = subprocess.Popen(args, env = env)
+        ret = proc.wait()
+        if ret != 0:
+            exit(ret)
```

Comment:
Use `sys.exit` instead of `exit`. The latter is something unrelated and misleading

https://docs.python.org/2/library/sys.html#sys.exit

https://docs.python.org/2/library/constants.html#exit